File: frontend/src/types/train.ts
export interface TrainSearchResult {
  id: string
  train_no: string
  train_code?: string | null
  train_type?: string | null
  service_date?: string | null
  origin_station_name?: string | null
  destination_station_name?: string | null
}

export interface TrainStop {
  id: string
  station_id: string
  station_name: string
  telecode?: string | null
  stop_order: number
  arrive_time?: string | null
  depart_time?: string | null
  day_offset: number
  dwell_minutes?: number | null
  mileage_km?: number | null
  longitude?: number | null
  latitude?: number | null
}

export interface TrainDetail extends TrainSearchResult {
  data_source: string
}

File: backend/app/schemas/train.py
from datetime import date, time
from uuid import UUID

from pydantic import BaseModel


class TrainSearchResult(BaseModel):
    id: UUID
    train_no: str
    train_code: str | None = None
    train_type: str | None = None
    service_date: date | None = None
    origin_station_name: str | None = None
    destination_station_name: str | None = None


class TrainDetail(TrainSearchResult):
    data_source: str


class TrainStopOut(BaseModel):
    id: UUID
    station_id: UUID
    station_name: str
    telecode: str | None = None
    stop_order: int
    arrive_time: time | None = None
    depart_time: time | None = None
    day_offset: int = 0
    dwell_minutes: int | None = None
    mileage_km: float | None = None
    longitude: float | None = None
    latitude: float | None = None
